paasta_tools/autoscaling/autoscaling_cluster_lib.py

```diff
@@ -49,518 +46,589 @@
+        slaves = [slave for slave in slaves_list if slave_pid_to_ip(slave['task_counts'].slave['pid']) in ips]
+        slave_ips = [slave_pid_to_ip(slave['task_counts'].slave['pid']) for slave in slaves]
+        instance_descriptions = self.describe_instances([], region=self.resource['region'],
+                                                        instance_filters=[{'Name': 'private-ip-address',
+                                                                           'Values': slave_ips}])
+        slave_instances = []
+        for slave in slaves:
+            ip = slave_pid_to_ip(slave['task_counts'].slave['pid'])
+            instances = self.get_instances_from_ip(ip, instance_descriptions)
+            if not instances:
+                log.warning("Couldn't find instance for ip {0}".format(ip))
+                continue
+            if len(instances) > 1:
+                log.error("Found more than one instance with the same private IP {0}. "
+                          "This should never happen")
+                continue
+            slave_instances.append({'ip': ip,
                                     'task_counts': slave['task_counts'],
                                     'hostname': slave['task_counts'].slave['hostname'],
                                     'id': slave['task_counts'].slave['id'],
                                     'pid': slave['task_counts'].slave['pid'],
                                     'instance_id': instances[0]['InstanceId']})
-    ret = []
-    instance_type_weights = get_instance_type_weights(sfr)
-    for slave in sfr_slave_instances:
-        instance_description = [instance_description for instance_description in sfr_instance_descriptions
-                                if instance_description['InstanceId'] == slave['instance_id']][0]
-        slave['instance_type'] = instance_description['InstanceType']
-        slave['instance_weight'] = instance_type_weights[slave['instance_type']]
-        ret.append(slave)
-    return ret
-
-
-def set_spot_fleet_request_capacity(sfr_id, capacity, dry_run, region=None):
-    """ AWS won't modify a request that is already modifying. This
-    function ensures we wait a few seconds in case we've just modified
-    a SFR"""
-    ec2_client = boto3.client('ec2', region_name=region)
-    with Timeout(seconds=AWS_SPOT_MODIFY_TIMEOUT):
-        try:
-            state = None
-            while True:
-                state = get_sfr(sfr_id, region=region)['SpotFleetRequestState']
-                if state == 'active':
+        ret = []
+        instance_type_weights = self.get_instance_type_weights()
+        for slave in slave_instances:
+            instance_description = [instance_description for instance_description in instance_descriptions
+                                    if instance_description['InstanceId'] == slave['instance_id']][0]
+            slave['instance_type'] = instance_description['InstanceType']
```

Comment:
As I mentioned before, can you define something that states what is actually in this? I still don't really like that there are other parts of the code base that are relying on this extra metadata being added without explicitly saying so.

```
class RichAWSInstace(object):
   """A regular AWS Instance object, with additional metadata."""

   def __init__(self, instance_type, instance_weight, instance):
        self.instance_type = instance_type
        self.instance_weight = instance_weight
        self.original_slave = slave
```

Then have this function return a list of of ``RichSlave`` objects.

Then everywhere that relies on this can say so with docs, variable names and type checking.